# Choosing a location for the borehole

## Loading and preparing data

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [2]:
df0 = pd.read_csv('/project/datasets/geo_data_0.csv')
df1 = pd.read_csv('/project//datasets/geo_data_1.csv')
df2 = pd.read_csv('/project//datasets/geo_data_2.csv')

In [3]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [4]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [5]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [6]:
df0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [7]:
df1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [8]:
df2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [9]:
df0.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

In [10]:
df1.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

In [11]:
df2.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

In [12]:
df0.duplicated(
).sum()

0

In [13]:
df1.duplicated().sum()

0

In [14]:
df2.duplicated().sum()

0

### Conclusion
Three data files have been examined. The files contain 100,000 rows and 5 columns each. There are no gaps or duplicates in the data, and scaling is not required.

## Model training and validation

In [15]:
def train_model(df):
    X = df.drop(['id', 'product'], axis=1)
    y = df['product']
    X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.25, random_state=12345)

    model = LinearRegression()
    model.fit(X_train, y_train)
    predicted_valid = model.predict(X_valid)
    rmse = mean_squared_error(y_valid, predicted_valid) ** 0.5
    y_and_pred = pd.DataFrame({'correct':y_valid,'predicted':predicted_valid})
    return model, rmse, y_and_pred 

In [16]:
model_0, rmse_0, cor_and_pred_0 = train_model(df0)

In [17]:
model_1, rmse_1, cor_and_pred_1 = train_model(df1)

In [18]:
model_2, rmse_2, cor_and_pred_2 = train_model(df2)

In [19]:
d = {'Model': ['model_0', 'model_1','model_2'], 
     'RMSE': [rmse_0, rmse_1, rmse_2],
     'Average stock of predicted raw materials':[cor_and_pred_0.predicted.mean(),cor_and_pred_1.predicted.mean(),
                                           cor_and_pred_2.predicted.mean()]}
model_results = pd.DataFrame(data=d)
model_results

Model       RMSE  Average stock of predicted raw materials
0  model_0  37.579422                                 92.592568
1  model_1   0.893099                                 68.728547
2  model_2  40.029709                                 94.965046

### Conclusion 
The data obtained were split into two samples at a ratio of 75 to 25. Models were trained using a linear regression algorithm for the three study areas. Average predicted raw stock and RMSE were calculated.  

## Preparation for profit calculation

Let's calculate the sufficient volume of raw materials for break-even development of a new well. Budget - 10 billion rubles, the number of wells - 200 pieces, the cost per barrel - 450 rubles.  

In [20]:
BUDGET = 10000000000
BOREHOLES = 200
BARREL_PRICE = 450

In [21]:
prod = BUDGET/BOREHOLES/BARREL_PRICE/1000
prod

111.11111111111111

For the development of the well not to be unprofitable it is necessary to produce more than 111 thousand barrels

In [22]:
cor_and_pred_0['correct'].mean()

92.07859674082927

In [23]:
cor_and_pred_1['correct'].mean()

68.72313602435997

In [24]:
cor_and_pred_2['correct'].mean()

94.88423280885438

Average reserve in regions: 
- region 0: 92.1 thousand barrels
- region 1: 68.7 thousand barrels
- region 2: 94.9 thousand barrels

### Conclusion.
In order for well development not to be unprofitable it is necessary to produce more than 111 thousand barrels. 
According to calculations based on historical data - average reserve in regions based on 25 thousand wells in each region:
- region 0: 92.1 thousand barrels
- region 1: 68.7 thousand barrels
- Region 2: 94.9 thousand barrels

Which is less than the required minimum of raw materials.

## Calculation of profits and risks 

Let's write a function to calculate the average profit, find the 95% confidence interval and estimate the risk of loss. Apply the `bootstrap' technique with 1000 samples. 

In [25]:
def revenue(target, probabilities, count):
    probs_sorted = probabilities.sort_values(ascending=False)
    selected = target[probs_sorted.index][:count]
    return BARREL_PRICE  * selected.sum() * 1000 - BUDGET


In [26]:
def bootstrap(df, region_name):
    state = np.random.RandomState(12345)
    values = []
    counter = 0
    for i in range(1000):
        target_subsample = df['correct'].sample(n = 500, replace = True, random_state = state)
        probs_subsample = df['predicted'][target_subsample.index] 
        profit = revenue(target_subsample, probs_subsample, 200)
        if profit < 0:
            counter += 1
        values.append(profit)
        
    values = pd.Series(values)
    risk = (values < 0).mean()
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)# < напишите код здесь>

    mean = values.mean()
    print('For region {:.0f}'.format(region_name))
    print("Average revenue: {:.2f} тыс. руб.".format(mean/1000))
    print("Boundaries of the 95% confidence interval:  от {:.2f} до {:.2f} тыс. руб.".format(lower/1000, upper/1000))
    print('Risk of loss: {:.1%}'.format(risk))

In [27]:
bootstrap(cor_and_pred_0, 0)

For region 0
Average revenue: 425938.53 тыс. руб.
Boundaries of the 95% confidence interval:  от -102090.09 до 947976.35 тыс. руб.
Risk of loss: 6.0%


In [28]:
bootstrap(cor_and_pred_1, 1)

For region 1
Average revenue: 515222.77 тыс. руб.
Boundaries of the 95% confidence interval:  от 68873.23 до 931547.59 тыс. руб.
Risk of loss: 1.0%


In [29]:
bootstrap(cor_and_pred_2, 2)

For region 2
Average revenue: 435008.36 тыс. руб.
Boundaries of the 95% confidence interval:  от -128880.55 до 969706.95 тыс. руб.
Risk of loss: 6.4%


## Conclusion

We calculated the possible average revenue from the development of deposits for regions 0, 1 and 2.


Region 0 - the highest possible revenue of 425938.53 thous. rubles. Risk of loss for this region is 6%. On the basis of calculated 95% confidential interval we may conclude that in 97.5% of cases losses at field development will be equal to 102090.09 thousand rubles, in 2.5% of cases it is possible to get profit to 947976.35 thousand rubles.

Region 1 - the highest possible revenue is 515222.77 thousand rubles. Risk of loss for this region is 1%. Based on calculated 95% confidence interval we can make a conclusion, that in 97.5% of cases profit in field development will be from 68873.23 thousand rubles, in 2.5% of cases it is possible to get profit up to 931547.59 thousand rubles.

Region 1 - the highest possible revenue is 435008.36 thousand rubles. Risk of loss for this region is 6.4%. On the basis of calculated 95% confidence interval it may be concluded that in 97.5% of cases losses at field development will be up to 128880.55 thousand rubles, in 2.5% of cases it is possible to get profit up to 969706.95 thousand rubles.

Based on the analysis the optimal region for development is region 1.
